src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -395,24 +465,63 @@ private bool TestStream(Stream stream)
             return false;
         }
 
-        private void RegisterHandler(Guid runtimeId, StreamHandler handler)
+        private void ProvideEndpointInfo(in IpcEndpointInfo endpointInfo)
         {
-            lock (_lock)
+            lock (_newEndpointInfoLock)
             {
-                if (!_cachedStreams.TryGetValue(runtimeId, out Stream stream))
+                IpcEndpointInfo? endpointInfoNullable = endpointInfo;
+                // Provide the endpoint info to the first handler that will accept it.
+                for (int i = 0; i < _newEndpointInfoHandlers.Count && endpointInfoNullable.HasValue; i++)
                 {
-                    throw new InvalidOperationException($"Runtime instance with identifier '{runtimeId}' is not registered.");
+                    EndpointInfoHandler handler = _newEndpointInfoHandlers[i];
+                    // Handler will return true if it has completed
+                    if (handler(ref endpointInfoNullable))
+                    {
+                        _newEndpointInfoHandlers.RemoveAt(i);
+                        i--;
+                    }
                 }
 
-                _handlers.Add(handler);
+                // If the endpoint info was not consumed, add it to the list
+                if (endpointInfoNullable.HasValue)
```

Comment:
nit: would it be easier to just return a ValuleTuple or a secondary ref bool in the handler, rather than nullifying the endpointInfo?

Reply:
I patterned this off of how the existing stream handling is done, but I can look at changing this to make it more clear what the contract is.

Reply:
It now returns a bool as an out parameter describing if the value was consumed.